airflow/jobs/scheduler_job_runner.py

```diff
@@ -1247,59 +1240,68 @@ def _create_dag_runs_dataset_triggered(
-                    DatasetEvent.timestamp <= exec_date,
-                ]
-                if previous_dag_run:
-                    dataset_event_filters.append(DatasetEvent.timestamp > previous_dag_run.execution_date)
-
-                dataset_events = session.scalars(
-                    select(DatasetEvent)
-                    .join(
-                        DagScheduleDatasetReference,
-                        DatasetEvent.dataset_id == DagScheduleDatasetReference.dataset_id,
-                    )
-                    .join(DatasetEvent.source_dag_run)
-                    .where(*dataset_event_filters)
-                ).all()
-
-                data_interval = dag.timetable.data_interval_for_events(exec_date, dataset_events)
-                run_id = dag.timetable.generate_run_id(
-                    run_type=DagRunType.DATASET_TRIGGERED,
-                    logical_date=exec_date,
-                    data_interval=data_interval,
-                    session=session,
-                    events=dataset_events,
-                )
+            )
+            if session.scalar(curr_date_query):  # dag already exists
+                continue
 
-                dag_run = dag.create_dagrun(
-                    run_id=run_id,
-                    run_type=DagRunType.DATASET_TRIGGERED,
-                    execution_date=exec_date,
-                    data_interval=data_interval,
-                    state=DagRunState.QUEUED,
-                    external_trigger=False,
-                    session=session,
-                    dag_hash=dag_hash,
-                    creating_job_id=self.job.id,
+            prev_exec_date = session.scalar(
+                select(DagRun.execution_date)
+                .where(
+                    DagRun.dag_id == dag.dag_id,
+                    DagRun.execution_date < exec_date,
+                    DagRun.run_type == DagRunType.DATASET_TRIGGERED,
                 )
-                Stats.incr("dataset.triggered_dagruns")
-                dag_run.consumed_dataset_events.extend(dataset_events)
-                session.execute(
-                    delete(DatasetDagRunQueue).where(DatasetDagRunQueue.target_dag_id == dag_run.dag_id)
+                .order_by(DagRun.execution_date.desc())
+                .limit(1)
+            )
+            dataset_event_filters = [
+                DagScheduleDatasetReference.dag_id == dag.dag_id,
+                DatasetEvent.timestamp <= exec_date,
+            ]
+            if prev_exec_date:
+                dataset_event_filters.append(DatasetEvent.timestamp > prev_exec_date)
+
+            dataset_events = session.scalars(
+                select(DatasetEvent)
+                .join(
+                    DagScheduleDatasetReference,
+                    DatasetEvent.dataset_id == DagScheduleDatasetReference.dataset_id,
                 )
+                .join(DatasetEvent.source_dag_run)
+                .where(*dataset_event_filters)
```

Comment:
I kind of wonder if we can eliminate the extra `prev_exec_date` query, but that can probably be a separate improvement later.